[IMP] web_editor: rework background-sizing option

sdegueldre opened the pull request:
Previously, background size and position for snippets was configured in
a modal, the user could choose a "focus point", which is rather
unintuitive. They could also choose to use the "contain" background
mode, with or without repeat, or a third option, where they could
customize everything, from the background size in both directions to
whether to repeat the background in only one or in both directions.

This commit streamlines the background-sizing options, only the "cover"
and "repeat pattern" options have been kept, as they cover almost every
use case. Their configuration has been moved into the options menu for
consistency with the rest of the editor, and the background's position
is now chosen by clicking on the corresponding menu option, and dragging
the background to the desired place.

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -769,7 +769,7 @@ var SnippetsMenu = Widget.extend({
                 return;
             }
-            if ($target.closest('#web_editor-top-edit, #oe_snippets, #oe_manipulators, .o_technical_modal, .oe_drop_zone').length) {
+            if ($target.closest('#web_editor-top-edit, #oe_snippets, #oe_manipulators, .o_technical_modal, .oe_drop_zone, .o_ignore_select').length) {
```
Your new UI should just be inside `#oe_manipulators` (which is supposed to contain all custom DOM that will be put over the editable DOM).

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1250,222 +1250,239 @@ registry.background = SnippetOption.extend({
+            },
+        };
+        // Fill inputs with current values.
+        this.$el.find('input').val((i) => this.backgroundSize[i === 0 ? 'x' : 'y']);
```
After reflexion, even if why not, I think it would be clearer in two lines ;) The day you have x,y,z,a,b,c,d,e, ok ;)

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1250,222 +1250,239 @@ registry.background = SnippetOption.extend({
+        // Rerender overlay content on window resized because background images change size
+        $(window).on('resize.bgposition', () => this._toggleOverlay(!this.$overlay.is('.d-none')));
+        // Rerender overlay on carousel slide because they sometimes take up more space and move elements around them
+        $(document).on('slid.bs.carousel.bgposition', () => this._toggleOverlay(!this.$overlay.is('.d-none')));
```
Share code by putting the anonymous function in a variable, or by binding both events on window at the same time

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1250,222 +1250,239 @@ registry.background = SnippetOption.extend({
+    destroy: function () {
+        this._super.apply(this, arguments);
+        $(window).off('resize.bgposition');
+        $(document).off('slid.bs.carousel.bgposition');
```
You can just unbind all events belonging to this class with `.off('.bgposition')`

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1250,222 +1250,239 @@ registry.background = SnippetOption.extend({
      */
     onFocus: function () {
         this.$el.toggleClass('d-none', this.$target.css('background-image') === 'none');
+        this.$el.find('we-input').toggleClass('d-none', this.$target.css('background-repeat') !== 'repeat');
```
I think we can do that in the `_updateUI` method instead now (maybe there should be an automatic `_updateUI` call on focus but that's another debate).

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1250,222 +1250,239 @@ registry.background = SnippetOption.extend({
+            case 'cover':
+                // Background size will be set to cover by css, values are saved in data-attributes
+                this.$target.css({backgroundSize: ''});
+                this.$target.toggleClass('o_bg_img_opt_repeat', false);
```
Use `addClass` and `removeClass`

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1250,222 +1250,239 @@ registry.background = SnippetOption.extend({
-                        'background-position': '',
-                    });
+    _initOverlay: function () {
+        this.$overlay = $(qweb.render('web_editor.background_position_overlay'));
```
There already is something in `this.$overlay`, name it differently :wink:

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,223 +1308,266 @@ registry.background = SnippetOption.extend({
+ * Handles the edition of snippets' background image position.
  */
-registry['background_position'] = SnippetOption.extend({
+registry.backgroundPosition = SnippetOption.extend({
```
Since you are renaming, rename with `BackgroundPosition`

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,223 +1308,266 @@ registry.background = SnippetOption.extend({
+        };
+        // Fill inputs with current values.
+        this.$el.find('.o_bg_width_input input').val(this.backgroundSize.x);
+        this.$el.find('.o_bg_height_input input').val(this.backgroundSize.y);
```
Let's do what we talked about:
- The background width in pixels (nothing being auto)
- Forget about saving the values when switching to cover... or do it internally in the option for the time of the edition (not writing the DOM)
- Enable your code so that the drag for the position is more coherent

sdegueldre replied:
Alright, it's done. Background size in pixels, using the `<we-input>`'s `data-set-style`, no longer saving size values, and dragging is synchronized between the mouse and background. Also rebased on top of your latest changes and everything seems to be working fine.

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,221 +1308,276 @@ registry.background = SnippetOption.extend({
+
+        this.$target.on('snippet-option-change', () => {
+          this._updateUI();
+          this.img.src = this._getSrcFromCssValue();
```
I see a misindentation here ;)

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,221 +1308,276 @@ registry.background = SnippetOption.extend({
         });
+
+        // Binding _onDocumentClicked so it can be attached and detached while still having access to methods
+        this._onDocumentClicked = this._onDocumentClicked.bind(this);
```
Use jquery namespace like you did for resize and slid and this won't be needed

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,221 +1308,276 @@ registry.background = SnippetOption.extend({
+          // Should probably take current input values instead of hardcoding defaults.
+          // Any clean way to do that?
+          this.$target.css('background-size', '100px 100px');
+          this.$target.css('background-position', '');
```
Again a problem of indentation here... :)

+ if/else contains the same stuff

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,221 +1308,276 @@ registry.background = SnippetOption.extend({
+        this.$manipulators = $('#oe_manipulators > :not(.o_background_position_overlay)').addClass('d-none');
+        // Create empty clone of $target with same display size, make it draggable and give it a tooltip.
+        this.$bgDragger = this.$target.clone().empty();
+        this.$bgDragger.on('mousedown', this._dragBackground.bind(this));
```
`_dragBackground` should be called `_onBackgroundDragStart` and positioned in the "handlers" section

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,221 +1308,276 @@ registry.background = SnippetOption.extend({
+        $(this.ownerDocument).one('mouseup', () => {
+            this.$bgDragger.css({cursor: ''});
+            $(this.ownerDocument).off('mousemove', drag);
+        });
```
I think jquery namespace would make the code more readable here too (you can just `off` without knowning the drag function and you are not forced to put it in a separated variable. If you don't agree, I'll accept :smile: 

However, put the `$(this.ownerDocument)` in a variable to not repeat it 3 times

sdegueldre replied:
I'm just quite used to locally scoped functions when dealing with event binding and unbinding, while I don't think either approach is more readable, I'll just follow what's already done elsewhere in the code base so it's less confusing.

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,221 +1308,276 @@ registry.background = SnippetOption.extend({
+        });
+        this.$backgroundOverlay.find('.o_btn_discard').on('click', () => this._toggleOverlay(false));
+
+        $('#oe_manipulators').append(this.$backgroundOverlay);
```
If possible, we avoid grepping the whole DOM for something that do not belong to this class. A less hacky way would be:
```javascript
this.$backgroundOverlay.insertAfter(this.$overlay);
```

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,221 +1308,276 @@ registry.background = SnippetOption.extend({
-                'background-size': this.previousState[1],
-                'background-position': this.previousState[2],
+    _dimensionOverlay: function () {
+        const $wrapwrap = $('#wrapwrap');
```
Also kinda hacky, searching for the highest ancestor of `this.$target` which is editable would be better I think (`:o_editable` selector ?) (this option is for all editors, not necessarily website... even if only used in website for now I think ?)

sdegueldre replied:
The thing is, we want to darken anything that is not part of the editor's interface, in our case that is the wrapwrap, and it's not `:o_editable`. The highest ancestor which is, is the `#wrap`, but even its parent is only the `<main>` so covering either of those would leave the header and footer unobscured.

An alternative would be to use something like bootstrap's modal backdrop, which would cover everything, including the editor's interface, which might not be a bad thing.

qsm-odoo replied:
@sdegueldre Mmh ok, I guess we will wait for the editor team to come up with a great way to retrieve that wrapwrap element. Meanwhile you still can't suppose it exists. Use `document.body` and use its `width` and `height` instead of `innerWidth` and `innerHeight` and you should achieve the same result.

sdegueldre replied:
As discussed, body isn't really an option because offset doesn't work with it, it has the wrong height, and parsing computedStyle paddings is brittle. Just noting it here for reference.

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,221 +1308,276 @@ registry.background = SnippetOption.extend({
+            return;
+        }
+        
+        this.$manipulators = $('#oe_manipulators > :not(.o_background_position_overlay)').addClass('d-none');
```
Should `trigger_up` something and handle this at snippetMenu level... or maybe is not this enough to trigger_up "deactivate_snippet" at the beginning then reactivate afterwards ?

sdegueldre replied:
Same problem as with deactivating a snippet while the carousel is sliding, the menu would show the "Blocks" section of the menu instead of staying on the snippet being currently modified which looks weird. Ideally this would indeed be handled by the snippetMenu, but with a different event to just hide the overlay of the current snippet, and another one to show it again. It just seemed like a lot of infrastructure work for something that may only be used by one snippet, but if you think it's worth it to add that feature to the snippetMenu directly, I will do that.

qsm-odoo replied:
Mmh, maybe not, I'll think about it when a second case appears. Meanwhile, you can still replace by `this.$overlay.siblings(':not(.o_background_position_overlay)')` to find those elements.

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,221 +1308,276 @@ registry.background = SnippetOption.extend({
+        this.$bgDragger.tooltip('show');
+
+        // Needs to be deferred or the click event that activated the overlay deactivates it as well.
+        _.defer(() => $(document).on('click', this._onDocumentClicked));
```
As we try to use ES6 now, we also try to use less underscorejs for nothing, so use `setTimeout` here :)

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1308,221 +1308,276 @@ registry.background = SnippetOption.extend({
+        return {
+            x: this.$target.outerWidth() - Math.round(renderRatio * this.img.naturalWidth),
+            y: this.$target.outerHeight() - Math.round(renderRatio * this.img.naturalHeight),
         }
```
Your linter should not be happy about this missing `;`

sdegueldre replied:
It seems my linter was throwing a fatal error on the newly added async function in the color picker because it wasn't configured for the right ES version, quite a few things flew under my radar. I was confused why it was no longer warning me about some stuff, problem solved.

qsm-odoo commented on addons/website/views/snippets.xml:
```diff
@@ -1224,10 +1224,18 @@
+            <we-button data-background-type="repeat-pattern">Repeat pattern</we-button>
+        </we-select>
+        <we-input class="o_bg_width_input" string="Background width" data-set-style="background-size" data-default-value="100" data-unit="px"/>
+        <we-input class="o_bg_height_input" string="Background height" data-set-style="background-size" data-default-value="100" data-unit="px"/>
```
default value should be auto for the height

sdegueldre replied:
At the moment my code doesn't handle `auto` values for background-size calculations. It seems kind of weird to me that auto would be supported at all here, auto preserves the background original's aspect ratio, but you can do the same thing with pixel values, and `auto auto` is the image's natural size which you can also just type into the inputs, so the use case is already covered with the numbers.

And then there's the confusion factor of why an input with a unit behind it supports a keyword (`auto px` doesn't mean anything in css) and why it doesn't support other keywords (like contain or initial) or typing in units. IMO these fields should either be unitless and support everything (which as we have discussed is brittle and might break in unexpected ways depending on browsers' standard compliance so it's a bad idea) or, and this is what I think we should do here, have a fixed unit (determined by the xml view as is already the case) and numeric inputs, not supporting any keyword at all.

This has the advantage that, being a native field, it is accessible by default, supports a min and max, has keyboard controls built-in (for which you have a custom handler right now) and supports steps, while only allowing numeric inputs.

sdegueldre replied:
At the moment, my background size calculation doesn't support `auto` . Since `auto` preserves the background's aspect ratio, the use case is covered by just inputting the corresponding pixel value, and `auto auto` keeps the image's natural dimensions so you can also just use those pixel values to cover that use case, and I could fill the inputs with the image's natural dimensions on startup.

And then there's the weirdness that comes from the fact that auto would be supported at all in a field with a unit, `auto px` doesn't mean anything in CSS. What's more, why would it support one keyword in particular but not others (like `contain` or `initial`), or even other units?

As discussed previously, supporting all units is not feasible, and I think that supporting keywords is a bad idea for the same reasons: lots of edge cases and pitfalls. Also, not supporting keywords would allow us to use `<input type="number"/>` instead, which has a good number of advantages, such as input validation, native keyboard controls (instead of needing a custom handler as you have now), support for a step, min and max.

In the cases where we do want to support auto, I think we should use a separate option (like we do here with cover)

qsm-odoo replied:
@sdegueldre 

Clearly, auto should be supported IMO and the system I implemented supported it. In fact the inputs in your task probably do not work because the default value has to be written with the unit. So "100px"... but in this case, it should really be `auto` IMO. If I upload a 48x33 pattern and choose 67px width, I clearly don't expect to have to put 67/(48/33) = 46.0625px in the second input to not have my image deformed (and most people (like 87%) don't want their image deformed).

> And then there's the weirdness that comes from the fact that auto would be supported at all in a field with a unit, auto px doesn't mean anything in CSS. What's more, why would it support one keyword in particular but not others (like contain or initial), or even other units?

-> We don't support the user writing "auto" just the user removing the value so that it default to "auto" and I think this is clear he can't enter words precisely because 'px' is written behind. I think it is clear for most users (like 92%) that when you don't enter anything, it is computed automatically.

> Also, not supporting keywords would allow us to use <input type="number"/> instead

(So as I explained we don't support keywords) but using number inputs is debatable. I personally really don't like the UI being bloated with up and down buttons for each input, and there is the fact it renders differently on all browsers which we try to avoid in Odoo. But overall: number inputs are not supported by Safari on iOS so those users won't have the up/down/min/max/step features I planned to implement.

@sbuhl What is your opinion about all of this?

sdegueldre replied:
Alright, in that case it's fine to only support auto for height I guess?

Putting the unit in the default value solved the input problem.

qsm-odoo replied:
There is a problem with setting `data-default-value="auto"` for both of them ?

sdegueldre replied:
Not really, I just thought that since most of the default images are quite big, they would cover the entire background and not make it obvious it's repeating, so for new users it might be a little confusing. But I guess it's most convenient for people who are actually trying to use the feature to just use the image's natural dimensions.

Added some code for that and put both defaults to auto.

qsm-odoo commented on addons/website/views/snippets.xml:
```diff
@@ -1224,10 +1224,18 @@
+        <we-input class="o_bg_width_input" string="Background width" data-set-style="background-size" data-default-value="100" data-unit="px"/>
+        <we-input class="o_bg_height_input" string="Background height" data-set-style="background-size" data-default-value="100" data-unit="px"/>
+        <we-button data-background-position-overlay="true" data-no-preview="true">
+            <i class="fa fa-fw fa-arrows"/> Background position
```
You can already remove the icon

sdegueldre commented:
> the inputs are not working fine on my side (only work when putting two values in them)... but I guess this is my fault? (strange because I tested the two inputs case with the commit I showed you earlier and it worked... I'll check deeper tomorrow).

I have the same issue, when the input contains its default value (either because it was set initially, or if it's manually set to that value) it just doesn't set the style properly. No idea why.

sdegueldre commented:
Alright, made the last requested changes, reviewed the comments to make sure they're up to date, cherry-picked back on top of your work, fixed the overlay position since the manipulators' container is now on top of the wrapwrap instead of below it, and moved some static css to scss files instead of using jquery. Tested and seems to be working fine.

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1357,221 +1356,295 @@ registry.background = SnippetOption.extend({
+            this.$target.css('background-size', '');
+        } else {
+            this.$target.css('background-size', 'auto');
+        }
```
I guess we can simply add the background-size: auto as default for the o_bg_img_opt_repeat repeat rule then. I'll do that.

qsm-odoo commented on addons/web_editor/static/src/scss/wysiwyg_snippets.scss:
```diff
@@ -819,6 +819,19 @@ body.editor_enable.editor_has_snippets {
+            .o_overlay_content {
+                cursor: grab;
+
+                .grabbing {
```
Even this one should start with `o_` and in fact, all should start with `o_we_` since it is in web_editor. I'll change that.

qsm-odoo commented on addons/website/views/snippets.xml:
```diff
@@ -1228,11 +1228,18 @@
-    <!-- Background Image -->
-    <div data-js="background_position"
-        data-selector="section, .parallax, .carousel-item"
-        data-exclude=".o_gallery .carousel-item">
```
I think that the selector and exclude should not have change here. I'll change that.

qsm-odoo commented:
Merged in parent task's branch :+1:
